controllers/config_helper.go

```diff
@@ -0,0 +1,195 @@
+package controllers
+
+import (
+	"context"
+
+	operatorv1alpha1 "github.com/gatekeeper/gatekeeper-operator/api/v1alpha1"
+	"github.com/go-logr/logr"
+	"github.com/open-policy-agent/gatekeeper/v3/apis/config/v1alpha1"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/wildcard"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	cacheRuntime "sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/metrics/server"
+)
+
+// Default config data
+func getDefaultConfig(namespace string) *v1alpha1.Config {
+	config := &v1alpha1.Config{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      "config",
+		},
+		Spec: v1alpha1.ConfigSpec{
+			Match: []v1alpha1.MatchEntry{
+				{
+					ExcludedNamespaces: []wildcard.Wildcard{
+						"kube-*", "multicluster-engine",
+						"hypershift", "hive", "rhacs-operator", "open-cluster-*", "openshift-*",
+					},
+					Processes: []string{
+						"webhook", "mutation-webhook",
+					},
+				},
+			},
+		},
+	}
+
+	return config
+}
+
+func createDefaultConfig(ctx context.Context, c client.Client, namespace string,
+	gatekeeper *operatorv1alpha1.Gatekeeper, scheme *runtime.Scheme,
+) error {
+	var config *v1alpha1.Config
+
+	// When it is DisableDefaultMatches = true
+	if gatekeeper.Spec.Config != nil && gatekeeper.Spec.Config.DisableDefaultMatches != nil &&
+		*gatekeeper.Spec.Config.DisableDefaultMatches {
+		config = &v1alpha1.Config{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "config",
+				Namespace: namespace,
+			},
+		}
+	} else {
+		config = getDefaultConfig(namespace)
+	}
+
+	matches := []v1alpha1.MatchEntry{}
+	if gatekeeper.Spec.Config != nil && gatekeeper.Spec.Config.Matches != nil {
+		// Will append gatekeeper.Spec.Config.Matches
+		matches = gatekeeper.Spec.Config.Matches
+	}
+
+	// Append matched from gatekeeper cr spec.config.matches
+	config.Spec.Match = append(config.Spec.Match, matches...)
+
+	err := c.Create(ctx, config)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Reset Config.Spec.Match and append the default exempt namespaces and
+// provided matched from gatekeeper CR
+func setExemptNamespaces(ctx context.Context,
+	c client.Client, existingConfig *v1alpha1.Config,
+	gatekeeper *operatorv1alpha1.Gatekeeper, scheme *runtime.Scheme, log logr.Logger,
+) error {
+	// gatekeeper.Spec.Config.Matches = [] should update config.match
+	// gatekeeper.Spec.Config.Matches = []  and  gatekeeper.Spec.Config.Matches == nil are diff
+	if gatekeeper.Spec.Config == nil || gatekeeper.Spec.Config.Matches == nil {
+		log.V(1).Info("The gatekeeper matches does not exist. Skip adding DefaultExemptNamespaces")
```

Comment:
This should only be the case *if* there is not an owner reference of the Gatekeeper operator on existingConfig and existingConfig has matches set. Otherwise, the Gatekeeper operator assumes ownership and sets the default exempt namespaces.

Reply:
If user already has config, then  user set `gatekeeper.Spec.Config = ["has-something "]`, we won't change the config file?

Reply:
I'm saying that if the user has not set `matches` manually on the existing Config object, then the default exempt namespaces should be set.

Here, you are checking if the user set anything in the Gatekeeper operator CR which is not what determines if the feature is enabled/disabled. It's enabled by default unless the user has manual configuration prior to Gatekeeper 3.14.0.